asapdiscovery-ml/asapdiscovery/ml/models.py

```diff
@@ -150,7 +158,253 @@ def pull(self, local_dir: Union[Path, str] = None) -> "LocalMLModelSpec":
+        return models
+
+    @property
+    def ensemble_size(self):
+        return len(self.models)
+
+    def pull(self, local_dir: Union[Path, str] = None) -> "LocalEnsembleMLModelSpec":
+        """
+        Pull ensemble model from S3
+
+        Parameters
+        ----------
+        local_dir : Union[Path, str], optional
+            Local directory to store model files, by default None, meaning they will be stored in pooch.os_cache
+
+        Returns
+        -------
+        List[LocalMLModelSpec]
+            List of local model specs
+        """
+        return LocalEnsembleMLModelSpec(
+            models=[model.pull(local_dir) for model in self.models],
+            **self.dict(exclude={"models"}),
+        )
+
+
+def _url_to_yaml(url: str) -> dict:
+    # Retrieve the file content from the URL
+    response = requests.get(url, allow_redirects=True)
+    response.raise_for_status()
+    # Convert bytes to string
+    try:
+        content = response.content.decode("utf-8")
+    except:  # noqa: E722
+        raise ValueError(f"Failed to decode content from {url}")
+    # Load the yaml
+    return yaml.safe_load(content)
+
+
+def check_mtenn_version_set_compatible(models: list[MLModelSpecBase]) -> bool:
+    """
+    Check that all models in the list are compatible with the same mtenn version
+    Bit hacky but should work for now.
+    """
+
+    try:
+        cur_version = Version.parse(mtenn.__version__)
+    except AttributeError:
+        warnings.warn(
+            "No mtenn version found. Assuming compatibility, but note "
+            "that this may be incorrect."
+        )
+        return True
+
+    for i in range(len(models)):
+        for j in range(i + 1, len(models)):
+            # ugly sorry
+            model1, model2 = models[i], models[j]
+            low_pin1 = (
+                model1.mtenn_lower_pin
+                if model1.mtenn_lower_pin
+                else Version.parse("0.0.0")
+            )
+            upper_pin1 = (
+                model1.mtenn_upper_pin
+                if model1.mtenn_upper_pin
+                else Version.parse("999.999.999")
+            )
+            low_pin2 = (
+                model2.mtenn_lower_pin
+                if model2.mtenn_lower_pin
+                else Version.parse("0.0.0")
+            )
+            upper_pin2 = (
+                model2.mtenn_upper_pin
+                if model2.mtenn_upper_pin
+                else Version.parse("999.999.999")
+            )
+
+            # check if versions are compatible with current mtenn version
+            m1_version_compat = low_pin1 <= cur_version < upper_pin1
+            m2_version_compat = low_pin2 <= cur_version < upper_pin2
+
+            if not m1_version_compat or not m2_version_compat:
+                return False
+
+            # now check m1 and m2 are cross compatible
+            if not (low_pin1 <= upper_pin2 and low_pin2 <= upper_pin1):
+                return False
+
+    return True
```

Comment:
this works so fine to leave it is as, but could also do something like gather all lower pins and upper pins, then just make sure the `mtenn` version is between the highest lower pin and lowest upper pin (I think this is equivalent)